tsdb/show_tag_values.go

```diff
@@ -0,0 +1,307 @@
+
+			row := &models.Row{
+				Name:    tvs.Key,
+				Columns: []string{"tagKey"},
+				Values:  make([][]interface{}, 0, lim-off),
+			}
+
+			for _, tv := range tvs.Values[off:lim] {
+				v := []interface{}{tv}
+				row.Values = append(row.Values, v)
+			}
+
+			out <- row
+		}
+	}()
+	return out
+}
+
+// Close closes the executor such that all resources are released. Once closed,
+// an executor may not be re-used.
+func (e *ShowTagValuesExecutor) close() {
+	if e != nil {
+		for _, m := range e.mappers {
+			m.Close()
+		}
+	}
+}
+
+// ShowTagValuesMapper is a mapper for collecting measurement names from a shard.
+type ShowTagValuesMapper struct {
+	remote    Mapper
+	shard     *Shard
+	stmt      *influxql.ShowTagValuesStatement
+	chunkSize int
+	state     interface{}
+}
+
+// NewShowTagValuesMapper returns a mapper for the given shard, which will return data for the meta statement.
+func NewShowTagValuesMapper(shard *Shard, stmt *influxql.ShowTagValuesStatement, chunkSize int) *ShowTagValuesMapper {
+	return &ShowTagValuesMapper{
+		shard:     shard,
+		stmt:      stmt,
+		chunkSize: chunkSize,
+	}
+}
+
+// TagValues represents measurement tag values.
+type TagValues struct {
+	Key    string   `json:"key"`
+	Values []string `json:"values"`
+}
+
+// TagsValues represents values for multiple tag keys.
+type TagsValues []*TagValues
+
+func (a TagsValues) Len() int           { return len(a) }
+func (a TagsValues) Less(i, j int) bool { return a[i].Key < a[j].Key }
+func (a TagsValues) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+// Size returns the total string length of tag key and values.
+func (a TagsValues) Size() int {
+	n := 0
+	for _, tv := range a {
+		n += len(tv.Key)
+		for _, v := range tv.Values {
+			n += len(v)
+		}
+	}
+	return n
+}
+
+// Open opens the mapper for use.
+func (m *ShowTagValuesMapper) Open() error {
+	if m.remote != nil {
+		return m.remote.Open()
+	}
+
+	// This can happen when a shard has been assigned to this node but we have not
+	// written to it so it may not exist yet.
+	if m.shard == nil {
+		return nil
+	}
+
+	sources := influxql.Sources{}
+
+	// Expand regex expressions in the FROM clause.
+	if m.stmt.Sources != nil {
+		var err error
+		sources, err = m.shard.index.ExpandSources(m.stmt.Sources)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Get measurements from sources in the statement if provided or database if not.
+	measurements, err := measurementsFromSourcesOrDB(m.shard.index, sources...)
+	if err != nil {
+		return err
+	}
+
+	// If a WHERE clause was specified, filter the measurements.
+	if m.stmt.Condition != nil {
+		var err error
+		whereMs, err := m.shard.index.measurementsByExpr(m.stmt.Condition)
+		if err != nil {
+			return err
+		}
+
+		sort.Sort(whereMs)
+
+		measurements = measurements.intersect(whereMs)
+	}
+
+	// Create a channel to send tag values on.
+	ch := make(chan TagsValues)
+	// Start a goroutine to send the names over the channel as needed.
+	go func() {
```

Comment:
Why do it this way? Could you not just iterate over the measurements in `NextChunk()`?